### PR TITLE
[KEYCLOAK-11707] Add documentation for the Elytron Credential Store p…

### PR DESCRIPTION
**Pull request by sguilhen**
…rovider

### CODE REVIEW

File: server_admin/topics/vault.adoc [hunk 1]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+
+For more detailed information on how to create/manage elytron credential stores, as well as how to mask keystore secrets,
+please refer to the Elytron documentation.
+
+NOTE: The `elytron-cs-keystore` vault provider has been implemented as a WildFly extension and as such is only available
+if the {project_name} server runs on WildFly/JBoss EAP.
+
+=== Key Resolvers
+
+All built-in providers support the configuration of one or more key resolvers. A key resolver is essentially implements
```
Comment (hmlnarik) [inline server_admin/topics/vault.adoc]:
```suggestion
All built-in providers support the configuration of one or more key resolvers. A key resolver essentially implements
```

File: server_admin/topics/vault.adoc [hunk 2]
```diff
@@ -24,7 +28,11 @@ To use a vault, a vault provider must be registered within {project_name}.
 It is possible to either use a built-in provider described below or
 implement your own provider. See the link:{developerguide_link}[{developerguide_name}] for more information.
 
-=== Kubernetes / OpenShift files plaintext vault provider
+NOTE: There is at most one vault provider active per {project_name} instance
+at any given time, and the vault provider in each instance within the cluster
+have to be configured consistently.
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: Singular, so it should be "has" rather than "have"

File: server_admin/topics/vault.adoc [hunk 3]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
-at any given time, and the vault provider in each instance within the cluster
-have to be configured consistently.
+=== Elytron Credential Store Vault Provider
+
+{project_name} also provides support for reading secrets stored in an Elytron credential store. The `elytron-cs-keystore`
+vault provider is capable of retrieving secrets from the keystore-based implementation of the credential store, which
+is also the default implementation provided by Elytron.
+
+This credential store is backed by a keystore (`JCEKS` is the default format, but it is possible to use other formats such as `PKCS12`)
+and users can create and manage the store contents using either the `elytron` subsystem in WildFly/JBoss EAP, or using the
```
Comment (andymunro) [inline server_admin/topics/vault.adoc:72]: should "WildFly/JBoss EAP" be {appserver_name}?

File: server_admin/topics/vault.adoc [hunk 4]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+`elytron-tool.sh` script.
+
+To use this provider, you have to declare the `elytron-cs-keystore` in the `keycloak-server` subsystem and set the location
+and master secret of the keystore that was created by Elytron. An example of the minimal configuration for the provider is shown bellow:
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: below (one L)

better format still is "...for the provider follows:"

File: server_admin/topics/vault.adoc [hunk 5]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+    <provider name="elytron-cs-keystore" enabled="true">
+        <properties>
+            <property name="location" value="${jboss.home.dir}/standalone/configuration/vault/credential-store.jceks" />
+            <property name="secret" value="secretpw1!"/>
+        </properties>
+    </provider>
+</spi>
+----
+
+If the underlying keystore has a format other than `JCEKS`, this has to be informed using the `keyStoreType`:
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: "this" should always precede a noun to make it clear what "this" is referring to.

File: server_admin/topics/vault.adoc [hunk 6]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+    <provider name="files-plaintext" enabled="true">
+        <properties>
+            <property name="dir" value="${jboss.home.dir}/standalone/configuration/vault/" />
+            <property name="keyResolvers" value="REALM_UNDERSCORE_KEY, KEY_ONLY"/>
+        </properties>
+    </provider>
+</spi>
+----
+
+The resolvers are executed in the same order they are declared in the configuration. For each resolver, the final entry
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: should be "... THAT they are declared.."

File: server_admin/topics/vault.adoc [hunk 7]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+name produced by the resolver that combines the realm with the vault key is used to search for the secret in the vault.
+If a secret is found, it is immediately returned. If not, the next resolver is used and this continues until a non-empty
+secret is found or all resolvers have been tried, in which case an empty secret is returned. In the example above, first
+the `REALM_UNDERSCORE_KEY` resolver is used. If an entry is found in the vault with the name it produces, it is returned.
+If not, then the `KEY_ONLY` resolver is used. Again, if an entry is found in the vault with the name it produces, it is
+returned. If not, an empty secret is returned since there are no more resolvers to be used.
+
+A list of the currently available resolvers is described bellow:
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: "...resolvers follows:"

File: server_admin/topics/vault.adoc [hunk 8]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+
+* `KEY_ONLY`: the realm name is ignored and the key from the vault expression is used as is.
+* `REALM_UNDERSCORE_KEY`: the realm and key are combined using an underscore `_` character. Occurrences of underscore in either the
+realm or key are escaped by another underscore character. So if the realm is called `master_realm` and the key is `smtp_key`, the
+combined key will be `master+++__+++realm_smtp+++__+++key`.
+* `REALM_FILESEPARATOR_KEY`: the realm and key are combined using the platform file separator character. This is useful in situaions
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: "situations"

File: server_admin/topics/vault.adoc [hunk 9]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+with the keys are grouped by realm using a directory structure.
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: ..the keys THAT are group...

File: server_admin/topics/vault.adoc [hunk 10]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+* `FACTORY_PROVIDED`: the realm and key are combined using the `VaultKeyResolver` that is provided by the vault provider factory. This
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: This combination allows...

File: server_admin/topics/vault.adoc [hunk 11]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+allows for the creation of a custom key resolver by extending an existing factory and implementing the `getFactoryResolver` method.
+
+If no resolver is configured for the built-in providers, the `REALM_UNDERSCORE_KEY` is selected by default.
+
+The `FACTORY_PROVIDED` resolver provides a hook that can be used to implement a custom resolver by extending the provider
```
Comment (andymunro) [inline server_admin/topics/vault.adoc:170]: returns not return

File: server_admin/topics/vault.adoc [hunk 12]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+        }
+
+        ...
+    }
+----
+
+The custom factory returns a key resolver that combines the realm and key with a triple `#` character. So an entry would look
+like `master_realm###smtp_key`, for example. This factory must then be installed just like any custom provider.
+
+Note that the custom factory must override both the `getFactoryResolver` and `getId` methods. The second is needed so that
```
Comment (andymunro) [inline server_admin/topics/vault.adoc]: The second method is needed...

File: server_development/topics/vault.adoc [hunk 13]
```diff
@@ -7,15 +7,22 @@ You can use a vault SPI from `org.keycloak.vault` package to write custom extens
 
 The built-in `files-plaintext` provider is an example of the implementation of this SPI. In general the following rules apply:
 
-* To prevent a secret from leaking across realms, you mey want to isolate or limit what secrets can be retrieved by a realm.
+* To prevent a secret from leaking across realms, you may want to isolate or limit what secrets can be retrieved by a realm.
```
Comment (andymunro) [inline server_development/topics/vault.adoc]: better is "...limit the secrets that can..."

File: server_admin/topics/vault.adoc [hunk 14]
```diff
@@ -54,6 +62,153 @@ Here is the equivalent configuration using CLI commands:
...
+   ...
+            <property name="secret" value="MASK-3u2HNQaMogJJ8VP7J6gRIl;12345678;321"/>
+   ...
+</spi>
+----
+
+For more detailed information on how to create/manage elytron credential stores, as well as how to mask keystore secrets,
+please refer to the Elytron documentation.
+
+NOTE: The `elytron-cs-keystore` vault provider has been implemented as a WildFly extension and as such is only available
```
Comment (andymunro) [inline server_admin/topics/vault.adoc:122]: should "WildFly/JBoss EAP" be {appserver_name}?

File: server_admin/topics/vault.adoc (hunk 3 above)
Comment (sguilhen) [reply server_admin/topics/vault.adoc:72]: I don't think so, we are specifically targeting this at WildFly/EAP and I believe {appserver_name} can yield different server names that we don't want listed here.

File: server_admin/topics/vault.adoc (hunk 4 above)
Comment (sguilhen) [reply server_admin/topics/vault.adoc]: Good suggestion, changing.

**Comment by sguilhen** [comment]
@andymunro thanks for the review, pushed the changes with your suggestions.

File: server_admin/topics/vault.adoc (hunk 14 above)
Comment (sguilhen) [reply server_admin/topics/vault.adoc:122]: I don't think so, see above.

**Review by hmlnarik — APPROVED** [review]

**Comment by hmlnarik** [comment]
Checks failed due to KEYCLOAK-12535, #808

**Pull request merged**